snuba/datasets/metrics_aggregate_processor.py

```diff
@@ -0,0 +1,137 @@
+        values = []
+        tags = message["tags"]
+        assert isinstance(tags, Mapping)
+        for key, value in sorted(tags.items()):
+            assert key.isdigit()
+            keys.append(int(key))
+            assert isinstance(value, int)
+            values.append(value)
+
+        processed = [
+            {
+                "org_id": ClickhouseInt(message["org_id"]),
+                "project_id": ClickhouseInt(message["project_id"]),
+                "metric_id": ClickhouseInt(message["metric_id"]),
+                "timestamp": ClickhouseUnguardedExpression(
+                    f"toStartOfInterval(toDateTime('{timestamp.isoformat()}'), toIntervalSecond({granularity}))"
+                ),
+                "tags.key": ClickhouseNumArray(keys),
+                "tags.value": ClickhouseNumArray(values),
+                **self._process_values(message),
+                "retention_days": ClickhouseInt(message["retention_days"]),
+                "partition": metadata.partition,
+                "offset": metadata.offset,
+                "granularity": ClickhouseInt(granularity),
+            }
+            for granularity in self.GRANULARITIES
+        ]
+        return AggregateInsertBatch(processed, None)
+
+
+class SetsAggregateProcessor(MetricsAggregateProcessor):
+    def _should_process(self, message: Mapping[str, Any]) -> bool:
+        return message["type"] is not None and message["type"] == "s"
+
+    def _process_values(self, message: Mapping[str, Any]) -> Mapping[str, Any]:
+        values = message["value"]
+        for v in values:
+            assert isinstance(v, int), "Illegal value in set. Int expected: {v}"
+        value_array = "[" + ",".join([str(v) for v in values]) + "]"
+
+        return {
+            "value": ClickhouseUnguardedExpression(
+                f"arrayReduce('uniqState', {value_array})"
+            )
+        }
+
+
+class CounterAggregateProcessor(MetricsAggregateProcessor):
+    def _should_process(self, message: Mapping[str, Any]) -> bool:
+        return message["type"] is not None and message["type"] == "c"
+
+    def _process_values(self, message: Mapping[str, Any]) -> Mapping[str, Any]:
+        value = message["value"]
+        assert isinstance(
+            value, (int, float)
+        ), "Illegal value for counter value. Int/Float expected {value}"
+
+        return {
+            "value": ClickhouseUnguardedExpression(
+                f"arrayReduce('sumState', [{value}])"
+            )
+        }
+
+
+class DistributionsAggregateProcessor(MetricsAggregateProcessor):
+    def _should_process(self, message: Mapping[str, Any]) -> bool:
+        return message["type"] is not None and message["type"] == "d"
+
+    def _process_values(self, message: Mapping[str, Any]) -> Mapping[str, Any]:
+        values = message["value"]
+        for v in values:
+            assert isinstance(
+                v, (int, float)
+            ), "Illegal value in set. Int expected: {v}"
```

Comment:
The assert is testing against both int and float but the assert message says that only Int is expected.